### ISSUE TITLE
Dev drive not mounted after reboot

### ISSUE THREAD

**Issue by rageshkrishna** [microsoft/devhome#1903]
### Dev Home version

0.601.297.0

### Windows build number

10.0.22631.0

### Steps to reproduce the bug

- Create a dev drive
- Store some code on it
- Reboot the machine

### Expected result

The dev drive should be mounted after the reboot

### Actual result

The dev drive is missing.

I am able to mount the VHD manually and it restores it to the right drive letter, but I'm unsure why this doesn't happen automatically. I have another machine where this works as expected and the mounted dev drive is available after reboots.

**Comment by niikoo** [comment]
+1

**Comment by jimmychanga** [comment]
I'm also experiencing this issue. Oddly enough I have another VHD with VMs on it that mounts automatically after a reboot.

**Comment by skanda890** [comment]
The Virtual Disk gets automatically removed when the computer reboots.

**Comment by skanda890** [comment]
I have done the same thing. I think I found the reason behind this:

Because the one which is still mounted is because something is stopping it from dismounting, but the other one is not getting mounted because, nothing is stopping it from dismounting. The one which is not getting dismounted should be a in a Storage Pool or something like that.

**Comment by seppestas** [comment]
I have the same issue, a colleague mentioned his dev drive remounts on reboot.

**Comment by JonathanBout** [comment]
Same problem over here! 

> I have the same issue. Made a scheduled task to mount it as a workaround.

@niikoo how did you do that exactly?

**Comment by niikoo** [comment]
This is how I set it up @JonathanBout

### Task Scheduler

**General**
Name: Auto-Mount DevDrive VHDX at Startup
When running the task, use the following user account: [I selected an account with administrative rights]
Selected `Run whether user is logged in or not` , `Do not store password` and `Run with highest privileges`.

**Triggers**
- At startup
- At log on

**Actions**
Action: `Start a program`
Program/script: `diskpart.exe`
Arguments: `/S "C:\devdrive.automount.diskpart"`

**Conditions**
Unchecked everything.

**Settings**
Checked the top 5 checkboxes. 
Set `Do not start a new instance` on the last dropdown.

### Create the diskpart file

Path: `"C:\devdrive.automount.diskpart"`
File contents: 
```
select vdisk file="C:\devDrive.vhdx"
attach vdisk
exit
```
Replace `"C:\devDrive.vhdx"` with the path to your DevDrive VHDX file.

**Comment by JonathanBout** [comment]
@niikoo thank you, it works great!

**Comment by skanda890** [comment]
> > The Virtual Disk gets automatically removed when the computer reboots.
> 
> That is not true, as I stated in my OP, I have 2 VHDs. One remains mounted after reboots, the dev drive does not.

@jimmychanga, does the one which stays still, in a storage pool.

**Comment by skanda890** [comment]
Close this if fixed.

**Comment by niikoo** [comment]
> Close this if fixed.

I don't see a workaround as a fix, @skanda890. Hope to get a proper fix that works out of the box.

**Comment by jimmychanga** [comment]
No, it is not. It is a NTFS formatted local vhd. The only difference between the two are that the DEV drive is formatted with ReFS.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by skanda890** [comment]
@jimmychanga, Thanks.

**Comment by mdsharpe** [comment]
Happening to me too, though intermittently. It seems random whether the dev drive is mounted on boot or not.

**Comment by AdamBraden** [comment]
For everyone that is experiencing this problem, can you open a Feedback Hub issue under Developer Platform->Dev Drive and collect traces? Then post the link here. 

Thanks,
Adam

**Comment by griffi-gh** [comment]
I unmounted my dev drive in order to optimize the vhdx file, when i mounted it back it won't automount after a reboot.  
Also, i can't create new dev drives, they just end up in RAW format and won't work unless i manually format them as NTFS (which defeats the whole point of dev drives)

**Comment by ChandKV** [comment]
Hi @griffi-gh, sorry to hear that. I wonder if the vhdx file was attached using diskmgmt.msc or other tools. If so, could you please detach the vhdx file using diskmgmt and retry attaching it using the **System\Storage\Disks & Volumes** Settings UX page? The behavior to auto-mount the vhdx files on reboot needs to be explicitly opted in and only the Settings UX for **Disks & Volumes** is updated to request the new behavior.

For the intermittent issue, thanks for reporting this. It will be more helpful to have the Feedback filed like @AdamBraden suggested as that would include the required logs. In this case, the System event log should have additional info on why the attempt to attach the vhd(x) failed.

**Comment by seppestas** [comment]
> The behavior to auto-mount the vhdx files on reboot needs to be explicitly opted in and only the Settings UX for **Disks & Volumes** is updated to request the new behavior.

This explains why it doesn't work for me. I mount the volume using explorer or from a shell script, mounting it through Settings does not work for me (probably due to some company managed policy).

**Comment by nzain** [comment]
Due to company-wide security policies dev drive creation via settings panel is not possible. For testing this feature, we used an elevated powershell with something along `new-VHD ... | Mount-VHD | ...` to create the dev drive. Its gone after a reboot.

**Comment by citelao** [comment]
> The behavior to auto-mount the vhdx files on reboot needs to be explicitly opted in and only the Settings UX for Disks & Volumes is updated to request the new behavior.

Indeed, there is a new [`_ATTACH_VIRTUAL_DISK_FLAG`](https://learn.microsoft.com/en-us/windows/win32/api/virtdisk/ne-virtdisk-attach_virtual_disk_flag) called `ATTACH_VIRTUAL_DISK_FLAG_AT_BOOT`, added sometime around the end of 2023. You can pass this argument to [`AttachVirtualDisk`](https://learn.microsoft.com/en-us/windows/win32/api/virtdisk/nf-virtdisk-attachvirtualdisk) to auto-mount VHDs. WIM and PowerShell have not been enlightened to do this.

I've written a sample C# app that will enable auto-mounting for a given VHD(X): [citelao/automount-vhd](https://github.com/citelao/automount-vhd). Most of the code is written by my friend @jazzdelightsme; I adapted it to CsWin32.

Note: I work for Microsoft, but I do not own these APIs.

**Comment by nzain** [comment]
Unfortunately, it seems this worked only once :-( Now, the script has no effect other than mounting the dev drive. Its gone after a reboot. However, I'm not sure if my company has some weird group policies that interfere somehow.

**Comment by JosiahBradley** [comment]
> Hi @griffi-gh, sorry to hear that. I wonder if the vhdx file was attached using diskmgmt.msc or other tools. If so, could you please detach the vhdx file using diskmgmt and retry attaching it using the **System\Storage\Disks & Volumes** Settings UX page? The behavior to auto-mount the vhdx files on reboot needs to be explicitly opted in and only the Settings UX for **Disks & Volumes** is updated to request the new behavior.
> 
> For the intermittent issue, thanks for reporting this. It will be more helpful to have the Feedback filed like @AdamBraden suggested as that would include the required logs. In this case, the System event log should have additional info on why the attempt to attach the vhd(x) failed.

Differentiating VHDX files is a dangerous act that leads to these exact bugs. There is no need to treat them differently, and if they truly are different then don't use the same underlying format, make a new one. I should be able to use my VHDX file between VMs, as a dev drive etc as that is what the format is for. Having one UI do one thing with hidden flags and another do another thing is not only inconsistent but how we get security issues etc. down the road as no one checks what the other system did. Respect the API as created.

I just started using dev drives and ran into this exact issue, of course after creating my VHDX via diskman. Forcing me to use the new clunky UI for the same tech (or worse preventing me from using PowerShell) is a scary future indeed.

**Comment by matt-gribben** [comment]
This has just started happening to me this week since rebuilding my machine / fresh Windows install

**Comment by adrastogi** [comment]
@matt-gribben would you mind filing a report from the Feedback Hub? You can log it under the 'Developer Platform -> Dev Drive' category and use the share option to include a link to it here. Thanks!

**Comment by matt-gribben** [comment]
@adrastogi https://aka.ms/AArjf34 there you go

**Comment by rickhoskins** [comment]
Your dev drive file needs to be on a permanent drive in order to be mounted on boot. If, like me, using a surface tablet, you have a SDHC card permanently in the machine, well, no good to put the dev drive file on that and want it to automount on boot.
Moved the file to the C: drive and mounted it, and it automounts on boot now. 

Sigh.

**Comment by AdamBraden** [comment]
@rickhoskins - putting a Dev Drive VHD on an SD card is not going to give you the perf benefits of Dev Drive compared to your SSD. Can you help me understand your desired scenario? If perf is not a major concern, you could just create a normal VHD and use a scheduled task to get it to mount - or put your content directly on the SD card. You could then use a folder exclusion to minimize Defender overhead (and do scheduled scans at appropriate times for the security threat/risk your project requires).

**Comment by JosiahBradley** [comment]
Is the only way to get traction on this the feedback hub? And if so why is there this Github repo if our bug reports come back with no response. The Dev Drive  concept is fairly DOA if you can't even mount things to it. I still have to manually mount my dev drive on every single reboot, which for me is often as I do not leave my computer on when I am not using it.

**Comment by AdamBraden** [comment]
@rickhoskins - thanks for the information on your scenario. The isolation and portability of a dev drive vhd is interesting to explore if more devs operate this way. 

@JosiahBradley - we are working with the file system team to fill in the gaps in the tools that support mounting and dismounting drives to ensure they support automount of a dev drive vhd. I cannot comment on a timeline at this time.

**Comment by Mukul1127** [comment]
> This is how I set it up @JonathanBout
> 
> ### Task Scheduler
> **General** Name: Auto-Mount DevDrive VHDX at Startup When running the task, use the following user account: [I selected an account with administrative rights] Selected `Run whether user is logged in or not` , `Do not store password` and `Run with highest privileges`.
> 
> **Triggers**
> 
> * At startup
> * At log on
> 
> **Actions** Action: `Start a program` Program/script: `diskpart.exe` Arguments: `/S "C:\devdrive.automount.diskpart"`
> 
> **Conditions** Unchecked everything.
> 
> **Settings** Checked the top 5 checkboxes. Set `Do not start a new instance` on the last dropdown.
> 
> ### Create the diskpart file
> Path: `"C:\devdrive.automount.diskpart"` File contents:
> 
> ```
> select vdisk file="C:\devDrive.vhdx"
> attach vdisk
> exit
> ```
> 
> Replace `"C:\devDrive.vhdx"` with the path to your DevDrive VHDX file.

Here's my Task that doesn't need a extra file:

### Task Scheduler
**General** Name: Auto-Mount DevDrive VHDX at Startup When running the task, use the following user account: [I selected an account with administrative rights] Selected `Run whether user is logged in or not` , `Do not store password` and `Run with highest privileges`.

**Triggers**

* At startup
* At log on

**Actions** Action: `Start a program` Program/script: `powershell` Arguments: `Mount-DiskImage -ImagePath "C:\DevDrive.vhdx"`

**Conditions** Unchecked everything.

**Settings** Checked the top 5 checkboxes. Set `Do not start a new instance` on the last dropdown. 

Replace `"C:\DevDrive.vhdx"` with the path to your DevDrive VHDX file.

**Comment by yashints** [comment]
This just happened to me too, I cloned the git repo @citelao shared and it worked, however, I am interested in the feature in Dev Home to mount an existing disk instead of always creating a new one

**Comment by fossbrandon** [comment]
I have the same issue. My event viewer provides an error log after that I can see after startup that shows the mounting fail. It shows a `Kernel-IO` error with the following info:

```
Failed to automatically attach a VHD during system startup.

          VHD name: C:\Users\foo\DevDrives\DevDrive.vhdxX̆偁瑰過
          Status: The request is not supported.
```

The VHD path it shows in the error appends random characters to the end of my path and I'm not sure if that's part of the problem or if it's something else. I'll go with one of the suggested temporary fixes for now, but it would be great to see this addressed so that the auto mounting is enabled and working by default.

**Comment by sidorov-alex** [comment]
I have exact same error in event viewer:

> Не удалось автоматически подключить виртуальный жесткий диск во время запуска системы.
> 
>           имя виртуального жесткого диска: E:\DevDrive.vhdx̀̃湍䅴+ㅢ
>           Состояние: Такой запрос не поддерживается.

VHDX File located at disk that is not removable.

Yesterday I began to investigate the problem on my own. I noticed that the Virtual Disk system service was disabled for some reason. So I set the startup type to automatic, rebooted, and viola - the disk mounted automatically.

BUT. Today after reboot the disk was not mounted again. Even though the Virtual Disk service is running.

**Comment by joncamp** [comment]
> I have the same issue. My event viewer provides an error log after that I can see after startup that shows the mounting fail. It shows a `Kernel-IO` error with the following info:
> 
> ```
> Failed to automatically attach a VHD during system startup.
> 
>           VHD name: C:\Users\foo\DevDrives\DevDrive.vhdxX̆偁瑰過
>           Status: The request is not supported.
> ```
> 
> The VHD path it shows in the error appends random characters to the end of my path and I'm not sure if that's part of the problem or if it's something else. I'll go with one of the suggested temporary fixes for now, but it would be great to see this addressed so that the auto mounting is enabled and working by default.

I am seeing variations on this. One is similar to yours, while the other two don't have the extra characters but do state "The request is not supported'`

```
Failed to automatically attach a VHD during system startup.

          VHD name: C:\DevDrive.vhdxç̃敓摓䆫઱
          Status: The request is not supported.
```

**Comment by carlosramosgonz** [comment]
Same problem here, and same message on boot on Event Viewer.

`No se pudo adjuntar automáticamente un VHD durante el inicio del sistema. 

          nombre de VHD: C:\DevDrive.vhdxů̃䵃扎ႄ뻭
          Estado: Solicitud no compatible.
`

When I do `reg query HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\AutoAttachVirtualDisks /s`:

`
HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\AutoAttachVirtualDisks\{aaf6f0dd-9029-4834-9bd5-d3eba12b79f2}
    Path    REG_SZ    C:\DevDrive.vhdx
`

**Comment by antonpiatek-eso** [comment]
Almost the same as all the above, but I get an additional kernel error in addition to the above.

```
Failed to automatically attach a VHD during system startup.

          VHD name: C:\DevDrive.vhdx
          Status: The specified VHD has an unexpected virtual disk identifier.
```

Also reg query shows two entries for the same dev drive

```
 reg query HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\AutoAttachVirtualDisks /s

HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\AutoAttachVirtualDisks\{8a9712ea-2627-4466-9083-0c74eeac6e1c}
    Path    REG_SZ    C:\DevDrive.vhdx

HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\AutoAttachVirtualDisks\{ea10e3f1-7f00-4e30-a682-db7215bf10fe}
    Path    REG_SZ    C:\DevDrive.vhdx
```

**Comment by emma-fosu** [comment]
Had the same problem with the dev drive not mounting after reboot. I fixed the problem by attaching the dev drive.

1.  Go to `Settings`>`Storage`>`Disks and Volumes`.
2.  Click on the `Attach VHD`.
3. Browse to the dev drive and select it.
4. Click on `Attach`.

Check if the dev drive is mounted and reboot to confirm if it has worked.

**Comment by patrick-stalph** [comment]
> Had the same problem with the dev drive not mounting after reboot. I fixed the problem by attaching the dev drive.
> 
> 1. Go to `Settings`>`Storage`>`Disks and Volumes`.
> 2. Click on the `Attach VHD`.
> 3. Browse to the dev drive and select it.
> 4. Click on `Attach`.
> 
> Check if the dev drive is mounted and reboot to confirm if it has worked.

The `Attach` button does nothing on my system. The popup where you browse for the dev drive file doesn't even close on pressing `Attach`.

**Comment by siewers** [comment]
I noticed another peculiarity. Whenever my laptop goes into hibernation, the DevDrive is still mounted, but after a reboot, it doesn't mount.
If I manually remount the drive and restart it, it will correctly mount.

So far, I think I have identified two cases where DevDrive doesn't remount:
1. when a Windows Update or driver has been installed
2. when the machine goes into hibernation and is restarted

**Comment by m-nagy** [comment]
I tried to run a Windows scheduled task to mount the disk using diskpart on startup, but it was completely unreliable. Despite using auto-unlock for the BitLocker drive, it didn't work every time.
I used a nice utility that auto-mounts the disk(s) at startup called [VhdAttach](https://github.com/medo64/VhdAttach)

**Comment by siewers** [comment]
@m-nagy Did you enable BitLocker on your DevDrive? If the VHD file is stored on a drive with BitLocker enabled, you don't need to add another layer of encryption. [Microsoft even says](https://learn.microsoft.com/en-us/windows/dev-drive/#will-a-vhd-created-for-use-as-a-dev-drive-be-encrypted-when-the-drive-storing-it-is-bitlocker-enabled) it's unnecessary.

**Comment by m-nagy** [comment]
@siewers  No, I didn't. As you said, I am using BitLocker only for the hosting volume.
